src/main/java/hudson/tasks/junit/TestResult.java

```diff
@@ -286,16 +288,30 @@ private void add(SuiteResult sr) {
         for (SuiteResult s : suites) {
             // JENKINS-12457: If a testsuite is distributed over multiple files, merge it into a single SuiteResult:
             if(s.getName().equals(sr.getName()) &&
-                    nullSafeEq(s.getId(),sr.getId()) &&
+                    eitherNullOrEq(s.getId(),sr.getId()) &&
                     nullSafeEq(s.getNodeId(),sr.getNodeId()) &&
                     nullSafeEq(s.getEnclosingBlocks(),sr.getEnclosingBlocks()) &&
                     nullSafeEq(s.getEnclosingBlockNames(),sr.getEnclosingBlockNames())) {
             
                 // However, a common problem is that people parse TEST-*.xml as well as TESTS-TestSuite.xml.
                 // In that case consider the result file as a duplicate and discard it.
                 // see http://jenkins.361315.n4.nabble.com/Problem-with-duplicate-build-execution-td371616.html for discussion.
-                if(strictEq(s.getTimestamp(),sr.getTimestamp())) {
-                    return;
+                if(strictEq(s.getTimestamp(),sr.getTimestamp()) &&
+                        s.getDuration() == sr.getDuration() &&
+                        s.getCases().size() == sr.getCases().size()) {
+                    
+                    // Rare edge case: testing framework writes one file per test case, but all same test suite...
+                    // *and* duration is not included, or so fast that first two files have 0.000 or 0.001.
+                    // When the second file is read, number of test cases will match... so compare the name.
+                    if (s.getCases().size() != 1 ||
+                            s.getCases().get(0).getFullName().equals(sr.getCases().get(0).getFullName())) {
```

Comment:
I'm no java developer but this seems to be very complex. Perhaps extract this logic to a private method and give it a self explanatory method name? That would also make the logic easier to test.

Regard the implementation I'm not sure adding comparisons for duration and filename will make this problem go away. I think there is two paths to take here.

1. calculate checksums for the report files and compare them
2. disable this logic with a configuration parameter like

**Jenkinsfile**
```groovy
junit 'tmp/junit.xml', disable_duplicate_report_check: true
```

I would like to see option 1 since that actually solves this bug. 
This implementation just makes it even harder to understand and debug.

Reply:
> 1. calculate checksums for the report files and compare them
This would break the original use case from years ago, which was separate files per testsuite, then a single file with all testsuites aggregated. The checksum wouldn't match, but the intention was to not add duplicate tests.

> 2. disable this logic with a configuration parameter like
I think more and more people will run into false positives where tests are disappearing and it's difficult to figure out why. If the logic does not work correctly in the default configuration, then maybe it should be removed... which is actually the direction that I've started moving towards. See comment that I will post shortly.